File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/BinaryTree/isSubPath_1367.java
package algo.tzashinorpu.SpecialSubject.BinaryTree;

import algo.tzashinorpu.SpecialSubject.ListNode;
import algo.tzashinorpu.SpecialSubject.TreeNode;

public class isSubPath_1367 {
    public boolean isSubPath(ListNode head, TreeNode root) {
        if (head == null) return true;
        if (root == null) return false;
        return dfs(head, root) || isSubPath(head, root.left) || isSubPath(head, root.right);
    }

    private boolean dfs(ListNode head, TreeNode root) {
        if (head == null) return true;
        if (root == null) return false;
        return head.val == root.val && (dfs(head.next, root.left) || dfs(head.next, root.right));
    }
}

File: Clion/ZeroTrac/Tree/isSubPath_1367.cpp
//
// Created by TzashiNorpu on 10/27/2022.
//
#include "tree.h"

using namespace Tree;

bool Solution::isSubPath(ListNode *head, TreeNode *root) {
  function<bool(ListNode *, TreeNode *)> dfs = [&](ListNode *list, TreeNode *node) -> bool {
      if (!list)return true;
      if (!node) return false;
      return list->val == node->val && (dfs(list->next, node->left) || dfs(list->next, node->right));
  };
  if (!head) return true;
  if (!root) return false;
  return dfs(head, root) || isSubPath(head, root->left) || isSubPath(head, root->right);
}